api/document/serializers/util.py

```diff
@@ -58,3 +59,25 @@ def list_to_internal_value(data: List[PrimitiveDict],
     # want.
     serializer = field_class(*args, **{'many': True, **kwargs})
     return serializer.to_internal_value(data)
+
+
+class SourcelineErrorMixin:
+    '''
+    A mixin for DRF Serializers that adds metadata about the
+    source line number that validation errors originated
+    from, if possible.
+
+    In order for this to work, the data passed to the
+    Serializer must have source line number metadata.
+    '''
+
+    def run_validation(self, data=empty):
+        try:
+            return super().run_validation(data)  # type: ignore
+        except ValidationError as e:
+            if (isinstance(e.detail, dict) and
+                    '_sourceline' not in e.detail and
+                    isinstance(data, dict) and
+                    isinstance(data.get('_sourceline'), int)):
```

Comment:
You've pointed to a few potential improvements in your comments, but I just wanted to flag this group of clauses as kinda hairy.

I'm not sure it's possible, but may be worth trying: if my mental model is right, there's a serializer instance associated with each object that gets deserialized. I could imagine the _state_ of those instances including the source line rather than tacking it on to the data. If that actually worked, we could reduce the number of corner cases that we need to check here.

Reply:
Hmm, I will think about if there's a way to simplify that gnarlyness a bit more, but right now I'm not sure if there is... Hmmmmmm.